### ISSUE TITLE
Unsigned integers?!

### ISSUE THREAD

**Issue by DevDevious** [go-jet/jet#71]
Given a database table mapped by go-jet to:

```
type Host struct {
	IDHost uint64 `sql:"primary_key"`
	Name string
	Addr string
	Port uint16
}
```

and

```

obj := model.Host {
    IDHost: 9980677793214571132,
    Name: "a-long-hash",
    Addr: "127.0.0.1",
    Port: 8080,
}

```
I should be able to do this

```

	stm := Host.INSERT(Host.AllColumns).
		MODEL(obj).
		ON_DUPLICATE_KEY_UPDATE(
			Host.IDHost.SET(UInt64(obj.IDHost)),
			Host.Name.SET(String(obj.Name)),
			Host.Addr.SET(String(obj.Addr)),
			Host.Port.SET(UInt16(obj.Port)),
			)
```

however jet-go has no `UIntXX` `IntegerExpression`. What am I missing?

**Comment by go-jet** [comment]
Currently there are no such wrappers. Just `Int`, which accepts `int64`. Feel free to open MR to add them. 
Or you can use `Int`: 
```go
Host.Port.SET(Int(int64(obj.Port))),
```

**Comment by DevDevious** [comment]
> Or you can use `Int`:
> 
> ```go
> Host.Port.SET(Int(int64(obj.Port))),
> ```

The above has drawbacks associated:
1) it will only be ok for primitives fitting an `int64`; 
2) sprinkling one's code with casts makes things unnecessarily verbose;

```go
Host.IDHost.SET(Int(int64(obj.IDHost)))
```

will overflow sooner or later, because IDHost is a 64 bit (unsigned) hash.

**Comment by DevDevious** [comment]
Will this work ?

``` go
// Int creates new integer literal
func Int(value int) IntegerExpression {
	numLiteral := &integerLiteralExpression{}

	numLiteral.literalExpressionImpl = *literal(value)
	numLiteral.literalExpressionImpl.Parent = numLiteral
	numLiteral.integerInterfaceImpl.parent = numLiteral

	return numLiteral
}

// Int8 creates new signed 8 bit integer literal
func Int8(value int8) IntegerExpression {
	numLiteral := &integerLiteralExpression{}

	numLiteral.literalExpressionImpl = *literal(value)
	numLiteral.literalExpressionImpl.Parent = numLiteral
	numLiteral.integerInterfaceImpl.parent = numLiteral

	return numLiteral
}

// Int16 creates new signed 16 bit integer literal
func Int16(value int16) IntegerExpression {
	numLiteral := &integerLiteralExpression{}

	numLiteral.literalExpressionImpl = *literal(value)
	numLiteral.literalExpressionImpl.Parent = numLiteral
	numLiteral.integerInterfaceImpl.parent = numLiteral

	return numLiteral
}

// Int32 creates new signed 32 bit integer literal
func Int32(value int32) IntegerExpression {
	numLiteral := &integerLiteralExpression{}

	numLiteral.literalExpressionImpl = *literal(value)
	numLiteral.literalExpressionImpl.Parent = numLiteral
	numLiteral.integerInterfaceImpl.parent = numLiteral

	return numLiteral
}

// Int64 creates new signed 64 bit integer literal
func Int64(value int64) IntegerExpression {
	numLiteral := &integerLiteralExpression{}

	numLiteral.literalExpressionImpl = *literal(value)
	numLiteral.literalExpressionImpl.Parent = numLiteral
	numLiteral.integerInterfaceImpl.parent = numLiteral

	return numLiteral
}


// UInt creates new unsigned integer literal
func UInt(value uint) IntegerExpression {
	numLiteral := &integerLiteralExpression{}

	numLiteral.literalExpressionImpl = *literal(value)
	numLiteral.literalExpressionImpl.Parent = numLiteral
	numLiteral.integerInterfaceImpl.parent = numLiteral

	return numLiteral
}

// UInt8 creates new unsigned 8 bit integer literal
func UInt8(value uint8) IntegerExpression {
	numLiteral := &integerLiteralExpression{}

	numLiteral.literalExpressionImpl = *literal(value)
	numLiteral.literalExpressionImpl.Parent = numLiteral
	numLiteral.integerInterfaceImpl.parent = numLiteral

	return numLiteral
}

// UInt16 creates new unsigned 16 bit integer literal
func UInt16(value uint16) IntegerExpression {
	numLiteral := &integerLiteralExpression{}

	numLiteral.literalExpressionImpl = *literal(value)
	numLiteral.literalExpressionImpl.Parent = numLiteral
	numLiteral.integerInterfaceImpl.parent = numLiteral

	return numLiteral
}

// UInt32 creates new unsigned 32 bit integer literal
func UInt32(value uint32) IntegerExpression {
	numLiteral := &integerLiteralExpression{}

	numLiteral.literalExpressionImpl = *literal(value)
	numLiteral.literalExpressionImpl.Parent = numLiteral
	numLiteral.integerInterfaceImpl.parent = numLiteral

	return numLiteral
}

// UInt64 creates new unsigned 64 bit integer literal
func UInt64(value uint64) IntegerExpression {
	numLiteral := &integerLiteralExpression{}

	numLiteral.literalExpressionImpl = *literal(value)
	numLiteral.literalExpressionImpl.Parent = numLiteral
	numLiteral.integerInterfaceImpl.parent = numLiteral

	return numLiteral
}
```

Of course the above could be simplified along the lines of

``` go
// UInt64 creates new unsigned 64 bit integer literal
func UInt64(value uint64) IntegerExpression {
	numLiteral := &integerLiteralExpression{}

	numLiteral.literalExpressionImpl = *literal(value)
	numLiteral.literalExpressionImpl.Parent = numLiteral
	numLiteral.integerInterfaceImpl.parent = numLiteral

	return numLiteral
}

// UInt32 creates new unsigned 32 bit integer literal
func UInt32(value uint32) IntegerExpression {
	return UInt64(uint64(value))
}
```

but the first approach should be lighter on memory.

**Comment by go-jet** [comment]
Maybe we could update existing `jet.Int` function:
`func Int(value int64) IntegerExpression {` => `func intLiteral(value interface{}) IntegerExpression {`,
and then support new integer types using `intLiteral`.
For instance:
```go
func UInt64(value uint64) IntegerExpression {
	return intLiteral(value)
}
```
Note that we can't change existing `mysql.Int` or `postgres.Int` because they are already part of a public api. 
This means we can't add integer wrapper for `int` and `uint`. This is fine because `int` and `uint` are machine dependent and generator will never generate model with this types.

**Comment by DevDevious** [comment]
> Maybe we could update existing `jet.Int` function:
> `func Int(value int64) IntegerExpression {` => `func intLiteral(value interface{}) IntegerExpression {`,
> and then support new integer types using `intLiteral`.

That is the way. :)

> For instance:
> 
> ```go
> func UInt64(value uint64) IntegerExpression {
> 	return intLiteral(value)
> }
> ```

Or maybe

``` go
func UInt64(value *uint64) IntegerExpression {
	return intLiteral(value)
}
```

so that we can SET nullable BIGINT unsigned columns easily like so

``` go
...
  .ON_DUPLICATE_KEY_UPDATE(MyTable.MyColumn.SET(UInt64(nil)))

```

What do you think ?

**Comment by go-jet** [comment]
It's a valid point. My concern is, it will complicate queries with non-null values.
For instance:
```go
MyTable.MyColumn.SET(UInt64(12))
```
would become:
```go
value := 12
MyTable.MyColumn.SET(UInt64(&value)) 
```
because we can't get address of a constant in Go.

There is a NULL literal that can be used for nullable columns, but it has to be first converted to exact type expression.
For instance:
```go
MyTable.MyColumn.SET(IntExp(NULL))
```

**Comment by DevDevious** [comment]
> My concern is, it will complicate queries with non-null values.

I agree. 

> There is a NULL literal that can be used for nullable columns, but it has to be first converted to exact type expression.
> For instance:
> 
> ```go
> MyTable.MyColumn.SET(IntExp(NULL))
> ```

I somehow missed that. `MyTable.MyColumn.SET(IntExp(NULL))` is exactly what I was looking for.

**Comment by DevDevious** [comment]
In conclusion, in `litera_expression.go` we would have:

```go
func intLiteral(value interface{}) IntegerExpression {
	numLiteral := &integerLiteralExpression{}

	numLiteral.literalExpressionImpl = *literal(value)

	numLiteral.literalExpressionImpl.Parent = numLiteral
	numLiteral.integerInterfaceImpl.parent = numLiteral

	return numLiteral
}

// Int creates a new 64 bit signed integer literal
func Int(value int64) IntegerExpression {
	return intLiteral(value)
}

// Int8 creates a new 8 bit signed integer literal
func Int8(value int8) IntegerExpression {
	return intLiteral(value)
}

// Int16 creates a new 16 bit signed integer literal
func Int16(value int16) IntegerExpression {
	return intLiteral(value)
}

// Int32 creates a new 32 bit signed integer literal
func Int32(value int32) IntegerExpression {
	return intLiteral(value)
}

// UInt8 creates a new 8 bit unsigned integer literal
func UInt8(value uint8) IntegerExpression {
	return intLiteral(value)
}

// UInt16 creates a new 16 bit unsigned integer literal
func UInt16(value uint16) IntegerExpression {
	return intLiteral(value)
}

// UInt32 creates a new 32 bit unsigned integer literal
func UInt32(value uint32) IntegerExpression {
	return intLiteral(value)
}

// UInt64 creates a new 64 bit unsigned integer literal
func UInt64(value uint64) IntegerExpression {
	return intLiteral(value)
}
```

And then in `literal.go`:

```go
// Int is constructor for 64 bit signed integer expressions literals.
var Int = jet.Int

// Int8 is constructor for 8 bit signed integer expressions literals.
var Int8 = jet.Int8

// Int16 is constructor for 16 bit signed integer expressions literals.
var Int16 = jet.Int16

// Int32 is constructor for 32 bit signed integer expressions literals.
var Int32 = jet.Int32

// Int64 is constructor for 64 bit signed integer expressions literals.
var Int64 = jet.Int

// UInt8 is constructor for 8 bit unsigned integer expressions literals.
var UInt8 = jet.UInt8

// UInt16 is constructor for 16 bit unsigned integer expressions literals.
var UInt16 = jet.UInt16

// UInt32 is constructor for 32 bit unsigned integer expressions literals.
var UInt32 = jet.UInt32

// UInt64 is constructor for 64 bit unsigned integer expressions literals.
var UInt64 = jet.UInt64
```

**Comment by go-jet** [comment]
LGTM.

**Comment by go-jet** [comment]
New integer literal constructor functions are now included in the [latests v2.5.0 release](https://github.com/go-jet/jet/releases/tag/v2.5.0).
